### DISCUSSION TITLE
Method to retrieve Collection Aggregate Summary details

### DISCUSSION THREAD

**Discussion by Debayanpaul1593** [Sunbird-Knowlg/Community#147] [Q&A] [answered]
Hi Sunbird Team,

I want to retrieve the activity aggregate, assessment aggregate and enrolment reconciliation data that is updated at the time of content/collection state update in my collection(course) hierarchy for every individual node.
There is one API `{{batch_service}}/v1/content/state/read` which returns the status of individual contents as a List. There is another API `/v1/collection/summary` which I am not able to run due to druid dependency.

I need to get the percentage completion at Course level and at Unit level. I want the data to include **lastreadcontentid** and **lastreadcontentstatus** datapoints from user_enrolments table as well. 
What is the best way to do this?

**Comment by AmiableAnil** [comment]
@Debayanpaul1593 you have to use the `{{batch_service}}/v1/content/state/read` to calculate the percentage calculation at any level of the course. 
We are returning the lastReadContentId in [UserEnrolment List API](https://staging.open-sunbird.org/api/course/v1/user/enrollment/list/{user-id})

For more detail on calculation on frontend @swayangjit can help you.

**Reply by Debayanpaul1593** [reply]
thanks @AmiableAnil I had one more query following this:
1. What are metrics and why is it being used in every job? Why are we incrementing metrics count? Where can I read more about it?
2. We were provided a document by the sunbird team. I am attaching the [confluence link](https://project-sunbird.atlassian.net/wiki/spaces/SBDES/pages/1520042020/Course+Infra+-+Async+Jobs+-+Implementation+Design#Course-Metrics-Aggregator%3A). Here there is a mention of **Course Metrics Aggregator** job. I was unable to find a job by this name in data-pipeline/lms-jobs or in knowledge-platform-jobs repo. The functionality seems similar to Activity Aggregate Updater job. Are these 2 the same?
Always thankful to the sunbird team for helping out with prompt replies

**Reply by AmiableAnil** [reply]
@Debayanpaul1593 
1. metrics is used for monitoring. If Grafana dashboard is setup so there we can see the lag, successful and failed event counts.
2. Activity Aggregate Updater job is the Course Metrics Aggregator job.

**Comment by Debayanpaul1593** [comment]
@AmiableAnil  How to do we fetch data stored in **user_activity_agg** table? It seems the **completedCount** value for any Course Unit is being stored in this table. `{{batch_service}}/v1/content/state/read` is retrieving data only from **user_content_consumption** table but  records for Course Units are not being stored here. Is there something I am missing?

**Reply by AmiableAnil** [reply]
@Debayanpaul1593 I'm not sure I understand your question. What are you looking for?

**Reply by Debayanpaul1593** [reply]
I have a published Course with the following hierarchy

- do_11384112739371417612 Course
   - do_11384112847498444815 Course Unit
      - do_11384112978137907219 pdf
      - do_113841130553425920112 video
   - do_11384112847502540817 Course Unit
      - do_113841129976094720110 pdf
      - do_113841130730700800113 video
   - do_11384112847468953613 Course Unit
      - do_113841130177912832111 pdf
      - do_113841134989213696114 video

When I am fetching content status for the course contents I am getting the response shown below:

```json
{
    "id": "api.content.state.read",
    "ver": "v1",
    "ts": "2023-08-01 16:40:43:121+0530",
    "params": {
        "resmsgid": null,
        "msgid": "ce9f665a-56e9-42fe-bb06-0c10e8442449",
        "err": null,
        "status": "success",
        "errmsg": null
    },
    "responseCode": "OK",
    "result": {
        "contentList": [
            {
                "lastAccessTime": "2023-07-18 17:46:40:108+0530",
                "contentId": "do_11384112978137907219",
                "batchId": "0138411744654213120",
                "completedCount": null,
                "progress": 100,
                "viewCount": null,
                "courseId": "do_11384112739371417612",
                "collectionId": "do_11384112739371417612",
                "lastCompletedTime": "2023-07-27 11:11:27:697+0530",
                "status": 2
            },
            {
                "lastAccessTime": "2023-07-20 22:02:19:564+0530",
                "contentId": "do_113841129976094720110",
                "batchId": "0138411744654213120",
                "completedCount": null,
                "progress": 100,
                "viewCount": null,
                "courseId": "do_11384112739371417612",
                "collectionId": "do_11384112739371417612",
                "lastCompletedTime": "2023-07-20 22:02:19:564+0530",
                "status": 2
            },
            {
                "lastAccessTime": "2023-07-27 11:43:30:160+0530",
                "contentId": "do_113841130177912832111",
                "batchId": "0138411744654213120",
                "completedCount": null,
                "progress": 20,
                "viewCount": null,
                "courseId": "do_11384112739371417612",
                "collectionId": "do_11384112739371417612",
                "lastCompletedTime": null,
                "status": 1
            },
            {
                "lastAccessTime": "2023-07-20 21:28:33:186+0530",
                "contentId": "do_113841130553425920112",
                "batchId": "0138411744654213120",
                "completedCount": null,
                "progress": 100,
                "viewCount": null,
                "courseId": "do_11384112739371417612",
                "collectionId": "do_11384112739371417612",
                "lastCompletedTime": "2023-07-20 21:49:18:518+0530",
                "status": 2
            },
            {
                "lastAccessTime": "2023-07-27 11:43:30:157+0530",
                "contentId": "do_113841130730700800113",
                "batchId": "0138411744654213120",
                "completedCount": null,
                "progress": 60,
                "viewCount": null,
                "courseId": "do_11384112739371417612",
                "collectionId": "do_11384112739371417612",
                "lastCompletedTime": null,
                "status": 1
            }
        ]
    }
}
```

It has the progress and status for the pdfs/videos but not for the Course Units. 
1. How do I get the progress and status of the Course Units and the Course as a whole? 
2. Is that computation done from the backend or needs to be done from the frontend? 
3. The Course Units are stored in **user_activity_agg** table with **aggregates_key** as completedCount and some numerical value as **aggregates_value**. How do I fetch and use this data?

**Reply by AmiableAnil** [reply]
@Debayanpaul1593 As I mentioned [here](https://github.com/orgs/Sunbird-Knowlg/discussions/147#discussioncomment-6548073) these calculation needs to be done from the frontend

**Reply by Debayanpaul1593** [reply]
ok got it. Thank you @AmiableAnil for all the help

**Comment by Debayanpaul1593** [comment]
[@swyangjit](https://github.com/swayangjit) can you give a brief overview of how to calculate percentage completion at Course Unit level for a given course from the frontend and the logic that you have used for the same?

**Reply by Krishnaj20** [reply]
@Debayanpaul1593 - I think you tagged the wrong Swayangjit. 

@swayangjit - Can you look into this query from Debayan?

**Reply by swayangjit** [reply, marked as answer]
@Debayanpaul1593  In mobile app we do an offline course calculation which you can find in below method
https://github.com/Sunbird-Ed/SunbirdEd-mobile-app/blob/85da3acf00a06e95000080ee5d40a90a83158f6c/src/app/enrolled-course-details-page/enrolled-course-details-page.ts#L1987

**Comment by Krishnaj20** [comment]
@Debayanpaul1593 - Let us know if your query is answered. If so, request you to mark the thread as answered.

**Reply by Debayanpaul1593** [reply]
Yes @Krishnaj20 query is answered. Thank you for your help.
